### ISSUE TITLE
Performance with isin function on large filter list

### ISSUE THREAD

**Issue by birdsarah** [dask/dask#4726]
I am opening a separate issue from #3198 as there seem to be two separate threads in that issue. Here's reproduce-able code:

```python
import dask.dataframe as dd
import pandas as pd

filter_list = list(range(2, 600000, 2))

for n in [10, 100, 1000]:
    df = dd.from_pandas(pd.DataFrame.from_dict({'A':[1, 2, 3, 4, 5] * 10_000}), npartitions=n)
    %time res = df.A.describe().compute()
```

    CPU times: user 51.9 ms, sys: 4.07 ms, total: 56 ms
    Wall time: 45.3 ms
    CPU times: user 231 ms, sys: 10.5 ms, total: 242 ms
    Wall time: 198 ms
    CPU times: user 2.32 s, sys: 167 ms, total: 2.49 s
    Wall time: 2.01 s

```python
for n in [10, 100, 1000]:
    df = dd.from_pandas(pd.DataFrame.from_dict({'A': [1, 2, 3, 4, 5] * 10_000}), npartitions=n)
    %time res = df[df['A'].isin(filter_list)].compute()
```

    CPU times: user 4.71 s, sys: 44.1 ms, total: 4.75 s
    Wall time: 4.73 s
    CPU times: user 45.2 s, sys: 313 ms, total: 45.5 s
    Wall time: 45.3 s
    CPU times: user 7min 46s, sys: 3.46 s, total: 7min 50s
    Wall time: 7min 48s

```python
for n in [10, 100, 1000, 10000]:
    df = dd.from_pandas(pd.DataFrame.from_dict({'A': [1, 2, 3, 4, 5] * 100_000}), npartitions=n) 
    mask = df.map_partitions(lambda x: x[x.A.isin(filter_list)])
    %time res = df.loc[mask.index].compute()
```

    CPU times: user 243 ms, sys: 12 ms, total: 255 ms
    Wall time: 200 ms
    CPU times: user 2.2 s, sys: 41.4 ms, total: 2.24 s
    Wall time: 1.86 s
    CPU times: user 20.6 s, sys: 214 ms, total: 20.8 s
    Wall time: 17.6 s
    CPU times: user 4min 47s, sys: 3.47 s, total: 4min 51s
    Wall time: 4min 11s

Direct use of `.isin` balloons in time to compute as the number of partitions increases. My observation from watching my system monitor, and as reported in #3198, is that the ballooning time comes from a single threaded processes that happens before the main processing kicks in. My guess is that this is part of building the task graph, but I don't know.

In my real world data processing, using map_partitions wasn't sufficient to get adequate performance and I needed to repartition.

**Comment by jcrist** [comment]
Thanks for the excellent issue report, I'll look into this this morning.

**Comment by jcrist** [comment]
Fixed by #4727. The issue was that the large list was placed as a literal in every task. This caused slowdowns in a few places:

- Serialization of the graph to workers was slower, as every task had a large piece of data in it.
- Evaluation of the graph was slower, as traversing a large list had to happen for each list
- Optimization of the graph was much slower (this was the single threaded slowdown you found above), as the large list was traversed multiple times for each partition during our optimization passes.

We now pull the data out to its own task, fixing all these issues.
